stan/math/prim/mat/vectorize/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,156 @@
+
+      /**
+       * Return the result of applying the function defined by the
+       * template parameter F to the specified scalars.
+       *
+       * @param x Scalar to which operation is applied.
+       * @param y Scalar to which operation is applied.
+       * @return Application of the function specified
+       * by F to the specified scalars.
+       */
+      static inline return_t apply(const T1& x, const T2& y) {
+        return F::fun(x, y);
+      }
+    };
+
+    template <typename F, typename T1, typename T2, int R1, int C1,
+              int R2, int C2>
+    struct apply_scalar_binary<F, Eigen::Matrix<T1, R1, C1>,
+                               Eigen::Matrix<T2, R2, C2> > {
+      /**
+       * Return type, which is calculated recursively as a matrix
+       * of the return type of the contained types T1 and T2.
+       */
+      typedef Eigen::Matrix<typename apply_scalar_binary<F, T1, T2>
+                                ::return_t, R1, C1> return_t;
+
+      /**
+       * Return the result of applying the function defined by the
+       * template parameter F to the specified matrix arguments.
+       *
+       * @param x Matrix to which operation is applied.
+       * @param y Matrix to which operation is applied.
+       * @return Componentwise application of the function specified
+       * by F to the specified matrices.
+       */
+      static inline return_t apply(const Eigen::Matrix<T1, R1, C1>& x,
+                                   const Eigen::Matrix<T2, R2, C2>& y) {
+        check_matching_dims("binary vectorization", "x", x, "y", y);
+        return_t result(x.rows(), x.cols());
+        for (int i = 0; i < x.size(); ++i)
+          result(i) = apply_scalar_binary<F, T1, T2>::apply(x(i), y(i));
+        return result;
+      }
+    };
+
+    template <typename F, typename T1, typename T2, int R1, int C1,
+              int R2, int C2>
+    struct apply_scalar_binary<F, Eigen::Block<Eigen::Matrix<T1, R1, C1> >,
+                               Eigen::Block<Eigen::Matrix<T2, R2, C2> > > {
+      /**
+       * Return type, which is calculated recursively as an Eigen::Matrix
+       * of the return type of the contained types T1 and T2.
+       */
+      typedef Eigen::Matrix<typename apply_scalar_binary<F, T1, T2>
+                                ::return_t, R1, C1> return_t;
+
+      /**
+       * Apply the function specified by F elementwise to the
+       * specified argument. This is defined recursively through this
+       * class by passing the arguments as matrices to the
+       * <code>Eigen::Matrix</code> specialization.
+       *
+       * @param x <code>Eigen::Block</code> argument container.
+       * @param y <code>Eigen::Block argument</code> container.
+       * @return Elementwise application of F to the elements of the
+       * containers.
+       */
+      static inline return_t apply(
+          const Eigen::Block<Eigen::Matrix<T1, R1, C1> >& x,
```

Comment:
You can do this more generally than explicitly coding for `Block`.  Check out:  https://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html
Specifically the section "How to write generic, but non-templated function?", which explains how to use references.  We haven't done this nearly as much as we should have elsewhere.

You can leave it like this if it works, but I'm thinking references will be more general if the template specialization types can be made to work.

Reply:
While I was trying to reduce the number of instantiations, I looked into using EigenBase<Derived>. However, the binary vectorization code wouldn't compile; it kept trying to use the default instantiation of T1 and T2 for Eigen::Matrix and failing. When I then tried it in a small test script, EigenBase<Derived> wasn't picked. I couldn't find any good documentation online so the only way I got it to work was using an Eigen::Block and Eigen::Matrix instantiation.

Reply:
OK, we've wrestled with this before and it's complicated.  We can leave it this way now and look to refactor it along with a bunch of other classes.